### PR TITLE
Feature/extension grant

### PR DESCRIPTION
**Pull request by DerLobi**
This adds support for extension grants as described in https://tools.ietf.org/html/rfc6749#section-4.5

I changed the percent-encoding of the HTTP body which was not working correctly. I based the new encoding on [AFNetworking's implementation](https://github.com/AFNetworking/AFNetworking/blob/master/AFNetworking/AFURLRequestSerialization.m#L64).

### CODE REVIEW

File: Heimdall/NSURLRequestExtensions.swift [hunk 1]
```diff
@@ -79,15 +79,35 @@ public extension NSMutableURLRequest {
...
+            HTTPBody = bodyString.dataUsingEncoding(NSUTF8StringEncoding)
         } else {
             HTTPBody = nil
         }
     }
+    
+    private func percentEscapedQueryStringKey(string: String) -> String {
+        let allowedCharacters = NSCharacterSet.URLQueryAllowedCharacterSet().mutableCopy() as! NSMutableCharacterSet
+        let charactersToLeaveUnescaped = NSCharacterSet(charactersInString:"[].")
+        let charactersToEscape = NSCharacterSet(charactersInString:":/?&=;+!@#$()',*")
```
Comment (felixjendrusch) [inline Heimdall/NSURLRequestExtensions.swift]: Please check https://github.com/Alamofire/Alamofire/blame/master/Source/ParameterEncoding.swift#L154, they have updated query string escaping. See also https://github.com/Alamofire/Alamofire/pull/370.

File: Heimdall/Heimdall.swift [hunk 2]
```diff
@@ -95,6 +95,19 @@ public class Heimdall {
...
     }
+    
+    /// Requests an access token with the given grant type URI and parameters
+    ///
+    /// **Note:** The completion closure may be invoked on any thread.
+    ///
+    /// :param: grantType The grant type URI of the extension grant
+    /// :param: parameters The required parameters for the external grant
+    /// :param: completion A callback to invoke when the request completed.
+    public func requestAccessToken(#grantType: String, parameters: [String: String], completion: Result<Void, NSError> -> ()) {
```
Comment (felixjendrusch) [inline Heimdall/Heimdall.swift]: I wonder whether we should expose the generic `requestAccessToken` method for future extensions.. opinions?

File: HeimdallTests/NSURLRequestExtensionsSpec.swift [hunk 3]
```diff
@@ -140,14 +140,28 @@ class NSMutableURLRequestExtensionsSpec: QuickSpec {
...
             context("when given parameters") {
-                it("sets the body with encoded parameyers") {
-                    request.setHTTPBody(parameters: [ "#key1": "%value1", "#key2": "%value2" ])
+                it("sets the body with encoded parameters") {
+                    
+                    let parameters = [
+                        "#key1": "%value1",
+                        "#key2": "%value2",
+                        "key3[]": "value3[]",
+                        "key4": ":/?&=;+!@#$()',*",
```
Comment (felixjendrusch) [inline HeimdallTests/NSURLRequestExtensionsSpec.swift]: !@#$%^&*()_+ :+1:

File: Heimdall/Heimdall.swift (hunk 2 above)
Comment (DerLobi) [inline Heimdall/Heimdall.swift]: I'd say no. If I read the RFC correctly the extension grant is the only way to implement a custom scheme. The grants we are still missing are "implicit" and "authorization code" and "client credentials". I'd add specific methods for these grants in the future.

File: HeimdallTests/HeimdallSpec.swift [hunk 4]
```diff
@@ -266,6 +266,192 @@ class HeimdallSpec: QuickSpec {
...
+                        }, withStubResponse: { request in
+                            return OHHTTPStubsResponse(data: NSData(contentsOfFile: self.bundle.pathForResource("authorize-invalid", ofType: "json")!), statusCode: 200, headers: [ "Content-Type": "application/json" ])
+                    })
+                    
+                    waitUntil { done in
+                        heimdall.requestAccessToken(grantType: "https://accounts.example.com/oauth/v2/foo/bar", parameters: ["provider": "fb", "code": "tops3cret"]) { result = $0; done() }
+                    }
+                }
+                
+                afterEach {
```
Comment (hffmnn) [inline HeimdallTests/HeimdallSpec.swift]: What is the difference between 301-303 vs 305-307?

Comment (DerLobi) [inline HeimdallTests/HeimdallSpec.swift]: no idea tbh, i copied the existing tests

Comment (DerLobi) [inline HeimdallTests/HeimdallSpec.swift]: fixed

File: Heimdall/NSURLRequestExtensions.swift [hunk 5]
```diff
@@ -76,18 +76,41 @@ public extension NSMutableURLRequest {
...
-            HTTPBody = "&".join(parts).dataUsingEncoding(NSUTF8StringEncoding)
+            var bodyString = join("&", components.map{"\($0)=\($1)"} as [String])
+            HTTPBody = bodyString.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)
         } else {
             HTTPBody = nil
         }
     }
+    
+    //Taken from https://github.com/Alamofire/Alamofire/blob/master/Source/ParameterEncoding.swift#L136
+    func queryComponents(key: String, _ value: AnyObject) -> [(String, String)] {
```
Comment (felixjendrusch) [inline Heimdall/NSURLRequestExtensions.swift]: Is this supposed to be `internal` or should it be `private`? Whitespace in the comment.

File: Heimdall/NSURLRequestExtensions.swift [hunk 6]
```diff
@@ -76,18 +76,41 @@ public extension NSMutableURLRequest {
...
-                let encodedName = name.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding)
-                let encodedValue = value.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding)
-                parts.append("\(encodedName!)=\(encodedValue!)")
+            var components: [(String, String)] = []
+            for (key, value) in parameters {
+                components += self.queryComponents(key, value)
             }
-
-            HTTPBody = "&".join(parts).dataUsingEncoding(NSUTF8StringEncoding)
+            var bodyString = join("&", components.map{"\($0)=\($1)"} as [String])
```
Comment (felixjendrusch) [inline Heimdall/NSURLRequestExtensions.swift]: Whitespace within the closure. Is the cast necessary?

File: Heimdall/NSURLRequestExtensions.swift [hunk 7]
```diff
@@ -76,18 +76,41 @@ public extension NSMutableURLRequest {
...
+            HTTPBody = bodyString.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)
```
Comment (felixjendrusch) [inline Heimdall/NSURLRequestExtensions.swift]: Isn't `allowLossyConversion: false` set by default?

File: Heimdall/NSURLRequestExtensions.swift [hunk 8]
```diff
@@ -76,18 +76,41 @@ public extension NSMutableURLRequest {
...
+    public func setHTTPBody(#parameters: [String: AnyObject]?) {
         if let parameters = parameters {
-            var parts = [String]()
-            for (name, value) in parameters {
-                let encodedName = name.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding)
-                let encodedValue = value.stringByAddingPercentEscapesUsingEncoding(NSUTF8StringEncoding)
-                parts.append("\(encodedName!)=\(encodedValue!)")
+            var components: [(String, String)] = []
+            for (key, value) in parameters {
+                components += self.queryComponents(key, value)
```
Comment (felixjendrusch) [inline Heimdall/NSURLRequestExtensions.swift]: I think the reference to `self` can be removed?!

File: Heimdall/OAuthAuthorizationGrant.swift [hunk 9]
```diff
@@ -14,12 +14,19 @@ public enum OAuthAuthorizationGrant {
     /// :param: refreshToken The refresh token.
     case RefreshToken(String)
 
+    /// An extension grant
+    ///
+    /// :param: grantType the grant type URI of the extension grant
+    /// :param: a dictionary of parameters
```
Comment (felixjendrusch) [inline Heimdall/OAuthAuthorizationGrant.swift]: Please add a parameter name for `[String: String]` to the documentation and use proper capitalization.

File: Heimdall/OAuthAuthorizationGrant.swift [hunk 10]
```diff
@@ -14,12 +14,19 @@ public enum OAuthAuthorizationGrant {
...
+    case Extension(String, [String: String])
+
     /// Returns the grant's parameters.
     ///
     /// Except for `grant_type`, parameters are specific to each grant:
     ///
     /// - `.ResourceOwnerPasswordCredentials`: `username`, `password`
     /// - `.Refresh`: `refresh_token`
+    /// - `.Extension`: `grantType`, `parameters`
```
Comment (felixjendrusch) [inline Heimdall/OAuthAuthorizationGrant.swift]: You don't actually set parameters with the names `grantType` or `parameters`.

Comment (DerLobi) [inline Heimdall/OAuthAuthorizationGrant.swift]: Well, I *kinda* do. Any idea how I can better describe it here?

**Comment by DerLobi** [comment]
@felixjendrusch thanks for the feedback, should be all fixed now

File: Heimdall/OAuthAuthorizationGrant.swift (hunk 10 above)
Comment (felixjendrusch) [inline Heimdall/OAuthAuthorizationGrant.swift]: My intention was to describe the parameters set in the returned array, e.g., `.Refresh` sets `grant_type` and `refresh_token`. So maybe something like: `Any parameters set in the extension's parameters array`.

**Comment by felixjendrusch** [comment]
Great work, thanks for updating the query escaping :+1:

**Comment by tibr** [comment]
Nice :+1:

**Pull request merged**